extensions/INTEL/INTEL_shader_integer_functions2.txt

```diff
@@ -0,0 +1,268 @@
+
+    2) How does countLeadingZeros differ from findMSB?
+
+    RESOLVED: countLeadingZeros is only defined for unsigned types, and it is
+    equivalent to 32-(findMSB(x)+1).  This corresponds the clz() function in
+    OpenCL and the LZD (leading zero detection) instruction on Intel GPUs.
+
+    3) How does countTrailingZeros differ from findLSB?
+
+    RESOLVED: countTrailingZeros is equivalent to max(genUType(findLSB(x)),
+    32).  This corresponds to the ctz() function in OpenCL.
+
+    4) Should 64-bit versions of countLeadingZeros and countTrailingZeros be
+    provided?
+
+    RESOLVED: NO.  OpenCL has 64-bit versions of clz() and ctz(), but OpenGL
+    does not have 64-bit versions of findMSB() or findLSB() even when
+    ARB_gpu_shader_int64 is supported.  The instructions used to implement
+    countLeadingZeros and countTrailingZeros do not natively support 64-bit
+    operands.
+
+    The implementation of 64-bit countLeadingZeros() would be 5 instructions,
+    and the implementation of 64-bit countTrailingZeros() would be 7
+    instructions.  Neither of these is better than an application developer
+    could achieve in GLSL:
+
+        uint countLeadingZeros(uint64_t value)
+        {
+            uvec2 v = unpackUint2x32(value);
+
+            return v.y == 0
+                ? 32 + countLeadingZeros(v.x) : countLeadingZeros(v.y);
+        }
+
+        uint countTrailingZeros(uint64_t value)
+        {
+            uvec2 v = unpackUint2x32(value);
+
+            return v.x == 0
+                ? 32 + countTrailingZeros(v.y) : countTrailingZeros(v.x);
+        }
+
+    5) Should 64-bit versions of the arithmetic functions be provided?
+
+    RESOLVED: NO.  Since recent generations of Intel GPUs have removed
+    hardware support for 64-bit integer arithmetic, there doesn't seem to be
+    much value in providing 64-bit arithmetic functions.
+
+    6) Should this extension include average()?
+
+    RESOLVED: YES.  average() corresponds to hadd() in OpenCL, and
+    averageRounded() corresponds to rhadd() in OpenCL.
+
+    averageRounded() corresponds to the AVG instruction on Intel GPUs.
+    average(), on the other hand, does not correspond to a single instruction.
+    The signed and unsigned versions may have slightly different
+    implementations depending on the specific GPU.  In the worst case, the
+    implementation is 4 instructions (e.g., averageRounded(x, y) - ((x ^ y) &
+    1)), and in the best case it is 3 instructions.
+
+Revision History
+
+    Rev  Date         Author    Changes
+    ---  -----------  --------  ---------------------------------------------
+      1  04-Sep-2018  idr       Initial version.
+      2  19-Sep-2018  idr       Add interactions with AMD_gpu_shader_int16.
+      3  22-Jan-2019  idr       Add interactions with EXT_shader_explicit_arithmetic_types.
+      4  14-Nov-2019  idr       Resolve issue #1 and issue #5.x
```

Comment:
There seems to be a stray "x" in the end of the line.

Reply:
Fixed locally.